File: 11287 - Pseudoprime Numbers.java
/*
*	User: Isanchez_Aguilar
*	Problem: UVA 11287 - Pseudoprime Numbers
*/
import java.io.OutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.PrintWriter;
import java.util.Scanner;
import java.math.BigInteger;

public class Main {
	public static void main(String[] args) {
		InputStream inputStream = System.in;
		OutputStream outputStream = System.out;
		Scanner in = new Scanner(inputStream);
		PrintWriter out = new PrintWriter(outputStream);
		UVA11287 solver = new UVA11287();
		solver.solve(1, in, out);
		out.close();
	}

	static class UVA11287 {
		public void solve(int testNumber, Scanner in, PrintWriter out) {
			BigInteger p;
			BigInteger a;

			while (!(p = in.nextBigInteger()).equals(BigInteger.ZERO)) {
				a = in.nextBigInteger();

				if (a.isProbablePrime(10) && p.isProbablePrime(10))
					out.println("no");
				else {
					if (a.modPow(p, p).equals(a))
						out.println("yes");
					else
						out.println("no");
				}
			}
		}

	}
}

File: 11287 - Pseudoprime Numbers.cpp
/*
*	User: Isanchez_Aguilar
*	Problem: UVA 11287 - Pseudoprime Numbers
*/
#include <bits/stdc++.h>

using namespace std;

inline long long modPow(long long a, long long p, long long m)
{
	a %= m;
	long long result = 1;

	while (p > 0)
	{
		if (p & 1)
			result = result * a % m;

		a = a * a % m;
		p >>= 1;
	}

	return result % m;
}

inline bool isPrime(long long n)
{
	long long sqrtN = sqrt(n);

	for (long long i = 2; i <= sqrtN; ++i)
	{
		if (n % i == 0)
			return false;
	}

	return true;
}

int main(void)
{
	ios_base::sync_with_stdio(0);
	cin.tie(0);

	long long p;
	long long a;
	while (cin >> p >> a and (p or a))
	{
		if (a == modPow(a, p, p) and not isPrime(p))
			cout << "yes\n";
		else
			cout << "no\n";
	}
	
	return 0;
}